src/inference/abc.jl

```diff
@@ -0,0 +1,279 @@
+using LinearAlgebra
+using GpABC
+
+struct GetData <: InferenceAlgorithm end
+
+struct DataState{T} <: DynamicPPL.AbstractSamplerState
+    observations::T
+end
+
+DynamicPPL.getspace(::Sampler{<:GetData}) = ()
+
+function DynamicPPL.assume(rng, spl::Sampler{<:GetData}, dist::Distribution, vn::DynamicPPL.VarName, vi)
+    r = rand(rng, dist)
+    push!(vi, vn, r, dist, spl)
+    return r, 0
+end
+
+function DynamicPPL.observe(spl::Sampler{<:GetData}, dist::Distribution, value, vi)
+    push!(spl.state.observations, value)
+    return 0
+end
+
+function DataState(model::DynamicPPL.Model)
+    return DataState([])
+end
+
+DynamicPPL.Sampler(alg::GetData, model::Model, s::DynamicPPL.Selector) = Sampler(alg, Dict{Symbol, Any}(), s, DataState(model))
+
+function get_data(model::Model)
+    # Get the data by executing the model once
+    spl = Sampler(GetData(), model, DynamicPPL.Selector())
+    var_info = Turing.VarInfo()
+    model(var_info, spl)
+
+    return [x for x in spl.state.observations] # trying to infer type
+end
+
+###################
+### AbstractABC ###
+###################
+abstract type AbstractABC <: InferenceAlgorithm end
+DynamicPPL.getspace(::Sampler{<:AbstractABC}) = ()
+
+function DynamicPPL.assume(rng, spl::Sampler{<:AbstractABC}, dist::Distribution, vn::VarName, vi)
+    # Sample from prior
+    r = rand(rng, dist)
+    push!(vi, vn, r, dist, spl)
+    return r, 0
+end
+
+function DynamicPPL.observe(spl::Sampler{<:AbstractABC}, dist::Distribution, value, vi)
+    # 1. Sample from likelihood
+    proposed = rand(dist)
+
+    # HACK: 2. Store proposal and observation (`value`) in sampler state so that we can later compute distance
```

Comment:
I can't find where observation is stored in sampler state. Am I missing anything?

Reply:
This is a left-over comment from before; I've removed this now :+1: